src/impl_owned_array.rs

```diff
@@ -658,6 +700,69 @@ impl<A, D> Array<A, D>
 
         Ok(())
     }
+
+    /// Reserve capacity to grow array along `axis` by at least `additional` elements.
+    ///
+    /// The axis should be in the range `Axis(` 0 .. *n* `)` where *n* is the
+    /// number of dimensions (axes) of the array.
+    ///
+    /// Existing elements of `array` are untouched and the backing storage is grown by
+    /// calling the underlying `reserve` method of the `OwnedRepr`.
+    ///
+    /// This is useful when pushing or appending repeatedly to an array to avoid multiple
+    /// allocations.
+    ///
+    /// ***Panics*** if the axis is out of bounds or if the new capacity would exceed `usize::MAX`.
+    ///
+    /// ***Errors*** with a shape error if the resultant capacity is larger than the addressable
+    /// bounds; that is, the product of non-zero axis lengths once `axis` has been extended by
+    /// `additional` exceeds `isize::MAX`.
+    ///
+    /// ```rust
+    /// use ndarray::{Array3, Axis};
+    /// let mut a = Array3::<i32>::zeros((0,2,4));
+    /// a.reserve(Axis(0), 1000).unwrap();
+    /// assert!(a.into_raw_vec().capacity() >= 2*4*1000);
+    /// ```
+    ///
+    pub fn reserve(&mut self, axis: Axis, additional: usize) -> Result<(), ShapeError>
+    where
+        D: RemoveAxis,
+    {
+        debug_assert!(axis.index() < self.ndim());
+        let self_dim = self.raw_dim();
+        let remaining_shape = self_dim.remove_axis(axis);
+
+        // Make sure added capacity doesn't overflow
+        debug_assert!(remaining_shape.size().checked_mul(additional).is_some());
+        let len_to_append = remaining_shape.size() * additional;
+
```

Comment:
I'm sorry for the late response here, but properties that are required for soundness (protects rules we necessarily uphold in our code and allows using `unsafe`) they **must be protected by assertions**, not debug assertions. The function must be a safe abstraction and have a sound result for any input parameter, for any regular compilation mode.

I haven't analyzed the code further, sometimes our preconditions are protected by other properties guaranteed by valid arrays and so on, so further thinking is needed but today I'm doing quick scans over PRs and won't delve into the details.

Reply:
Ah I see. Looking back since it's been a while, I think you're right that this check should be stronger. The call to `dimension::size_of_shape_checked` would already be catching cases where the overall size overflows but the new size of the growing dimension is still < `usize::MAX`, but it would miss the case of an overflow when calculating the new length of that growing dimension.

Since it would need to be evaluated anyway in an assert, I think just assigning from the `checked_mul` or returning a `ShapeError::Overflow` makes the most sense here (see most recent commit). Technically the addition to the growing dimension (impl_owned_array.rs:740 in that commit) could still overflow if another dimension had length 0, but in that case we would be reserving 0 bytes so I don't think that's a problem.

Reply:
thanks